Parse float: Bellerophon algorithm from "How to Read Floating Point Numbers Accurately"

tanishiking opened the pull request:
Implement bellerophon algorithm from "How to Read Floating Point Numbers Accurately" by William D. Clinger

TODO:
- [x] corner cases
  - [x] overflow
  - [x] underflow (subnormal / negative infinity)
- [x] Parse double
- [x] Parse hex (should be in another PR)
  - Math.pow and Long.parseLong are required
- [x] more test cases
  - maybe we wanna copy test cases from somewhere

---

This commit implements a Bellerophon algorithm from
"How to Read Floating Point Numbers Accurately" by William D. Clinger.

The Bellerophon algorithm efficiently finds the best binary
approximation for a decimal number `f * 10^e`.
While an iterative approach like AlgorithmM (defined in the same paper)
can find the best approximation by repeatedly using slow
BigInteger arithmetic, Bellerophon provides non-iterative approach.

The basic idea is, instead of directly computing the best approximation
for `f * 10^e` it uses higher-precision (p-bit (64-bit) significand)
floating-point numbers to approximate `f` and `10^e` separately.
Then multiplies these approximations and analyzes the
error when multiplying and rounding the result to the target
n-bit (n-bit significand) precision.

In some cases, the calculated approximation isn't the best, and in that case
the algorithm falls back to AlgorithmR.
This algorithm takes a good appoximation as a starting point,
and iteratively finds the best approximation.
Since the calculated approximation by Bellerophon is very close to the best
it converges in just a few steps.

For extreme values that would overflow to infinity or underflow to a
subnormal number or zero, neither the Bellerophon nor AlgorithmR does work.
In these cases, the algorithm falls back to the original, slower AlgorithmM
with the modification mentioned in section 8 of the paper.
> With IEEE arithmetic, for example, a denormalized result
> may be required. Denormalized results can be generate
> by a modified form of AlgorithmM that terminates immediately
> when the minimum exponent is reached.

In future, we might want to migrate to Eisel-Lemire algorithm?

tanishiking committed:
Wasm friendly parse float implementation

This commit implements a Bellerophon algorithm from
"How to Read Floating Point Numbers Accurately" by William D. Clinger.

The Bellerophon algorithm efficiently finds the best binary
approximation for a decimal number `f * 10^e`.
While an iterative approach like AlgorithmM (defined in the same paper)
can find the best approximation by repeatedly using slow
BigInteger arithmetic, Bellerophon provides non-iterative approach.

The basic idea is, instead of directly computing the best approximation
for `f * 10^e` it uses higher-precision (p-bit (64-bit) significand)
floating-point numbers to approximate `f` and `10^e` separately.
Then multiplies these approximations and analyzes the
error when multiplying and rounding the result to the target
n-bit (n-bit significand) precision.

In some cases, the calculated approximation isn't the best, and in that case
the algorithm falls back to AlgorithmR.
This algorithm takes a good appoximation as a starting point,
and iteratively finds the best approximation.
Since the calculated approximation by Bellerophon is very close to the best
it converges in just a few steps.

For extreme values that would overflow to infinity or underflow to a
subnormal number or zero, neither the Bellerophon nor AlgorithmR does work.
In these cases, the algorithm falls back to the original, slower AlgorithmM
with the modification mentioned in section 8 of the paper.
> With IEEE arithmetic, for example, a denormalized result
> may be required. Denormalized results can be generate
> by a modified form of AlgorithmM that terminates immediately
> when the minimum exponent is reached.

In future, we might want to migrate to Eisel-Lemire algorithm?

sjrd commented on javalib/src/main/scala/java/lang/Float.scala:
```diff
@@ -109,35 +110,45 @@ object Float {
 
-  def parseFloat(s: String): scala.Float = {
-    import Utils._
+  private[this] val parseFloatImpl = linkTimeIf[ParseFloatRegExpImpl](isWebAssembly) {
```
This will be needed elsewhere. Perhaps define it once and for all in an `object` in `java.util.regex.*`?

sjrd commented on javalib/src/main/scala/java/math/BigInteger.scala:
```diff
@@ -768,7 +772,11 @@ class BigInteger extends Number with Comparable[BigInteger] {
   override def toString(): String =
-    Conversion.toDecimalScaledString(this)
+    linkTimeIf(targetPureWasm) {
+      "" // TODO: Integer.toUnsignedString
```
If all you need is `Integer.toUnsignedString(i: Int)`, you can trivially implement that one as `Long.toString(Integer.toUnsignedLong(i))`. Those two methods are already implemented.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Tables.scala:
```diff
@@ -0,0 +1,72 @@
+   *  5^e < 2^64 <=> e < log5(2^64) <=> e < 27.563299716697156...
+   */
+  val smallPowerOfTens = Array[FloatingPoint](
+    FloatingPoint.normalized(new BigInteger("8000000000000000", 16), -63), // 0
```
It seems all the constants here are unsigned `Long` values. Store them as such, and convert them to `BigInteger` without having to parse a string?

This can be done with a small helper like
```scala
def ulongToBigInt(x: Long): BigInteger =
  if (x >= 0L) BigInteger.valueOf(x)
  else BigInteger.valueOf(x & ~Long.MinValue).setBit(63)
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Constants.scala:
```diff
@@ -0,0 +1,37 @@
+private[lang] object Constants {
+
+  final val ExtendedSigBits = 64
+  final val ExtendedMaxSig = BigInteger.TWO.pow(ExtendedSigBits)
```
```suggestion
  final val ExtendedMaxSig = BigInteger.ONE.shiftLeft(ExtendedSigBits)
```
or even
```suggestion
  final val ExtendedMaxSig = BigInteger.ZERO.setBit(ExtendedSigBits)
```
?

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,122 @@
+import Constants._
+
+/** DIY floating point number with 64-bit significand bits */
+private[dec2flt] class FloatingPoint private (val f: BigInteger, val e: Int) {
```
Does it always stay within 64 significant bits? If yes, why not use a `Long` (interpreted as unsigned) instead of a `BigInteger`?

tanishiking replied:
You're right. When multiplying these 64-bit floating-point numbers, an intermediate integer is up to 128 bits, but this could be represented by two `Long`s, and the result is immediately normalized back into the 64-bit range. So, it seems we can avoid using `BigInteger` here :+1:

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -244,7 +295,13 @@ object Double {
+      linkTimeIf(targetPureWasm) {
+        // java.lang.Long.parseLong may fail with NumberFormatException
+        // for a large input.
+        new BigInteger(s, radix).doubleValue()
```
When parsing `truncatedMantissaStr`, you should be able to use `parseUnsignedLong`. The largest possible string for `truncatedMantissaStr` is `"f" * 15 + "1"`, which correctly (though barely) parses to `-15L`.

When parsing `binaryExpStr`, if the string has more than 11 chars (10 digits and 1 sign), you can saturate it to `Int.MinValue`/`Int.MaxValue` (depending on the sign) without parsing it. If it has at most 11 chars, you can correctly parse it with `parseLong` (and then convert to `Int` with saturation; not wrapping).

tanishiking replied:
Ah, right, since `maxPrecisionChars = 15` it always fit in unsigned 64 bit integer 👍

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -244,7 +295,13 @@ object Double {
+        js.Dynamic.global.parseInt(s, radix).asInstanceOf[scala.Double]
+      }
 
     val mantissa = nativeParseInt(truncatedMantissaStr, 16)
```
So, concretely, I would leave `nativeParseInt` as being `nativeJSParseInt` and do
```suggestion
    val mantissa = linkTimeIf(targetPureWasm) {
      val mantissaLong = Long.parseUnsignedLong(truncatedMantissaStr, 16)
      // convert unsigned long to double
      (mantissaLong >>> 32).toDouble * (1L << 32) + (mantissaLong & 0xffffffffL).toDouble
    } {
      nativeJSParseInt(truncatedMantissaStr, 16)
    }
```
and below, for the computation of `binaryExp` (which does not show up in the diff so I can't comment on it directly):
```scala
    val binaryExp = linkTimeIf(targetPureWasm) {
      if (binaryExpStr.length() > 11) {
        if (binaryExpStr.charAt(0) = '-') Int.MinValue
        else Int.MaxValue
      } else {
        val binaryExpLong = Long.parseLong(binaryExpStr)
        if (binaryExpLong > Int.MaxValue.toLong) Int.MaxValue
        else if (binaryExpLong < Int.MinValue.toLong) Int.MinValue
        else binaryExpLong.toInt
      }
    } {
      val binaryExpDouble = nativeParseInt(binaryExpStr, 10)
      binaryExpDouble.toInt // caps to [MinValue, MaxValue]
    }
```

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -13,10 +13,12 @@
 package java.lang
 
 import java.lang.constant.{Constable, ConstantDesc}
+import java.math.BigInteger
```
Consider importing that only in the relevant methods, since it is a large dependency that we shouldn't use lightly.

tanishiking replied:
The input mantissa for Bellerophon still needs to be parsed as a BigInteger.
Other dependencies should be eliminated once `Math.scalb` implementation is merged.

sjrd replied:
I think you can move this one inside `parseDoubleWasm` now.

sjrd commented on javalib/src/main/scala/java/math/Conversion.scala:
```diff
@@ -167,7 +169,12 @@ private[math] object Conversion {
-      val absStr = Integer.toUnsignedString(digits(0))
+      // TODO: Integer.toUnsignedString hasn't yet implemented in Wasm
+      val absStr = linkTimeIf(targetPureWasm) {
+          java.lang.Long.toString(Integer.toUnsignedLong(digits(0)))
```
You could push that one step further into `Integer.toUnsignedString`.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -156,7 +153,6 @@ class DoubleTest {
       testFull("-87654.321", -87654.321)
       testFull("+.3f", 0.3)
 
-      // Hex notation, with exactly the output of toHexString()
```
Lost comment here :)

tanishiking commented:
@sjrd Thank you for comments! Once https://github.com/scala-js/scala-js/pull/5251 is merged, I'm planning to rebase on it.
Could you take a quick look especially on hex parsing? (we might want to do a serious code review on Bellorophon implementation when we upstream it?)

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -86,19 +88,28 @@ object Double {
-      "$")
-
-  private[this] lazy val doubleStrHexPat = new js.RegExp(
+  private[this] lazy val doubleStrPatStr =
```
I think there is no reason left to extract the strings in separate vals. It seems we can revert to the old structure, only with `RegExpImpl.impl.compile` instead of `new js.RegExp`.

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -108,29 +119,69 @@ object Double {
+      val e: Int = Integer.parseInt(exponentStr) - fractionalPartStr.length()
+      bellerophonDouble(f, e)
+    } else {
+      throw new NumberFormatException(s"""For input string: "$s"""")
```
This is dead code, isn't it? If the regex matches, we can't reach this point.

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -108,29 +119,69 @@ object Double {
-      js.Dynamic.global.parseFloat(undefOrForceGet[String](groups(1))).asInstanceOf[scala.Double]
+    import RegExpImpl.impl
+    val groups = impl.exec(doubleStrPat, s)
+    if (impl.matches(groups))
```
```suggestion
    if (impl.matches(groups)) {
```

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -108,29 +119,69 @@ object Double {
+        js.Dynamic.global.parseFloat(impl.get(groups, 1)).asInstanceOf[scala.Double]
+      }
     else
       parseDoubleSlowPath(s)
```
```suggestion
    } else {
      parseDoubleSlowPath(s)
    }
```

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -108,29 +119,69 @@ object Double {
+      scala.Double.PositiveInfinity
+    } else if (impl.exists(groups, 5)) {
+      // Decimal notation
+      val fullNumberStr = impl.get(groups, 5)
```
Unused?

sjrd commented on javalib/src/main/scala/java/lang/Double.scala:
```diff
@@ -108,29 +119,69 @@ object Double {
+      val fractionalPartStr =
+        impl.getOrElse(groups, 7, "") + impl.getOrElse(groups, 8, "")
+      val exponentStr = impl.getOrElse(groups, 9, "0")
+      val exponentStrWithoutSign = impl.getOrElse(groups, 10, "0")
```
Unused?

tanishiking replied:
Right, removed this, and unnecessary group from regex as well.

sjrd commented on javalib/src/main/scala/java/lang/Float.scala:
```diff
@@ -85,7 +86,7 @@ object Float {
   @inline def valueOf(s: String): Float = valueOf(parseFloat(s))
 
-  private[this] lazy val parseFloatRegExp = new js.RegExp(
+  private[this] lazy val parseFloatRegExpStr =
```
Same thing here. Go back to
```suggestion
  private[this] lazy val parseFloatRegExp = RegExpImpl.impl.compile(
```
?

sjrd commented on javalib/src/main/scala/java/lang/Integer.scala:
```diff
@@ -554,7 +554,13 @@ object Integer {
   }
 
-  @inline def toUnsignedString(i: scala.Int): String = toUnsignedString(i, 10)
+  @inline def toUnsignedString(i: scala.Int): String =
```
```suggestion
  @inline def toUnsignedString(i: scala.Int): String = {
```

sjrd commented on javalib/src/main/scala/java/lang/Long.scala:
```diff
@@ -241,10 +241,11 @@ object Long {
   // Must be called only with a valid radix
   @inline def parseLongPlatform(s: String, radix: Int): scala.Long = {
-    if (LinkingInfo.isWebAssembly)
+    LinkingInfo.linkTimeIf (LinkingInfo.isWebAssembly) {
```
```suggestion
    LinkingInfo.linkTimeIf(LinkingInfo.isWebAssembly) {
```

sjrd commented on javalib/src/main/scala/java/lang/Long.scala:
```diff
@@ -360,10 +361,11 @@ object Long {
   // Must be called only with a valid radix
   @inline def parseUnsignedLongPlatform(s: String, radix: Int): scala.Long = {
-    if (LinkingInfo.isWebAssembly)
+    LinkingInfo.linkTimeIf (LinkingInfo.isWebAssembly) {
```
```suggestion
    LinkingInfo.linkTimeIf(LinkingInfo.isWebAssembly) {
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+  def bellerophonDouble(f: BigInteger, e: Int): scala.Double =
+    bellerophon(f, e, Binary64)
+
+ /** An implementation of Bellerophon algorithm from
```
This entire comment is missing one space at the beginning of the line.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+  *  > when the minimum exponent is reached.
+  */
+  private def bellerophon(f: BigInteger, e: Int, fmt: FloatingPointFormat): fmt.Repr = {
+    // Fast path: when f < 2^sigbits (2^53 for Double), and |e| < ceil(log5(2^sigbits)),
```
Use
```scala
/* one
 * two
 */
```
for multiline comments.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+    } else if (f.compareTo(BigInteger.valueOf(fmt.MaxSig)) < 0 && e < 0 && -e < fmt.CeilLog5OfMaxSig) {
+      fmt.div(fmt.fromLong(f.longValueExact()), fmt.powerOfTen(-e))
+
+      // Instead of computing it directly, bellerophon calculate the best approximation by:
```
What does this comment refer to? Comments usually refer to the thing *after* them, but this comment is at the end of a block.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+    // As `e` is an integer, the condition is equivalent to e < ceil(log5(2^sigbits)).
+    if (f.compareTo(BigInteger.valueOf(fmt.MaxSig)) < 0 && e >= 0 && e < fmt.CeilLog5OfMaxSig) {
+      fmt.mul(fmt.fromLong(f.longValueExact()), fmt.powerOfTen(e))
+    } else if (f.compareTo(BigInteger.valueOf(fmt.MaxSig)) < 0 && e < 0 && -e < fmt.CeilLog5OfMaxSig) {
```
`f.compareTo(BigInteger.valueOf(fmt.MaxSig))` is done twice. Do it only once before the `if`s.

Also, consider storing `BigInteger.valueOf(fmt.MaxSig)` as an instance property of `fmt`.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+    } else if (f.compareTo(ExtendedMaxSig) >= 0 && (e < 0 || e >= TableSize)) {
+      multiplyAndTest(f, e, 4, fmt)
+    } else {
+      throw new Error(s"f: ${f.longValue()}, e: $e")
```
Is this supposed to be dead code? If yes, throw an `AssertionError`?

sjrd replied:
Hum, it does look like it's dead code. Consider reformulating the 4 tests above as:
```scala
val slop = if (f.compareTo(ExtendedMaxSig) < 0) {
  if (e >= 0 && e < TableSize) 0
  else 3
} else {
  if (e >= 0 && e < TableSize) 1
  else 4
}
multiplyAndTest(f, e, slop, fmt)
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+        val lowBits = z.f & ((1L << (ExtendedSigBits - fmt.SigBits)) - 1)
+        val diff = Math.abs(lowBits - (1L << (ExtendedSigBits - fmt.SigBits - 1)))
+        if (diff <= slop) {
+          // The value is too close to a rounding boundary use a safe backup algorithm.
```
```suggestion
          // The value is too close to a rounding boundary. Use a safe backup algorithm.
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      smallPowerOfTens(e)
+    } else {
+      val q = e / TableSize
+      val r = e % TableSize
```
I think if you use `Math.floorDiv` here, you won't need the path `e < 0`:
```scala
val q = Math.floorDiv(e, TableSize)
val r = e - q * TableSize
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      } else if (q.compareTo(MinSig) >= 0 && q.compareTo(MaxSig) < 0) {
+        val z = fmt.toIEEE754(FloatingPoint(q, k))
+        roundByRemainder(q, rem, v, z)
+      } else if (q.compareTo(MinSig) < 0) {
```
This comparison is repeated as well.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+        val z = fmt.toIEEE754(FloatingPoint(q, k))
+        roundByRemainder(q, rem, v, z)
+      } else if (q.compareTo(MinSig) < 0) {
+        loop(u.multiply(BigInteger.TWO), v, k - 1)
```
```suggestion
        loop(u.shiftLeft(1), v, k - 1)
```
?

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      } else if (q.compareTo(MinSig) < 0) {
+        loop(u.multiply(BigInteger.TWO), v, k - 1)
+      } else { // x > 2^m
+        loop(u, v.multiply(BigInteger.TWO), k + 1)
```
```suggestion
        loop(u, v.shiftLeft(1), k + 1)
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+        z: fmt.Repr): fmt.Repr = {
+      val vMinusR = v.subtract(rem)
+      if (rem.compareTo(vMinusR) < 0) z
+      else if (rem.compareTo(vMinusR) > 0) fmt.nextFloat(z)
```
Another repeated comparison.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      val vMinusR = v.subtract(rem)
+      if (rem.compareTo(vMinusR) < 0) z
+      else if (rem.compareTo(vMinusR) > 0) fmt.nextFloat(z)
+      else if (q.remainder(BigInteger.TWO).intValue() == 0) z // even
```
```suggestion
      else if (!q.testBit(0)) z // even
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      if (q.compareTo(MinSig) < 0) { // mantissa is small enough
+        roundByRemainder(q, rem, v, fmt.reinterpretBits(q))
+      } else {
+        throw new Error("Shouldn't reach here.")
```
```suggestion
        throw new AssertionError("Shouldn't reach here.")
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      val (x, y) = {
+        if (e >= 0) {
+          if (exp >= 0)
+            (f.multiply(BigInteger.TEN.pow(e)), mantissa.multiply(BigInteger.ZERO.setBit(exp)))
```
```suggestion
            (f.multiply(BigInteger.TEN.pow(e)), mantissa.shiftLeft(exp))
```
also in the other 3 below.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+        }
+      }
+      val d = x.subtract(y)
+      val d2 = BigInteger.TWO.multiply(mantissa).multiply(d.abs())
```
```suggestion
      val d2 = mantissa.shiftLeft(1).multiply(d.abs())
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+
+      if (d2.compareTo(y) < 0) { // d2 < y
+        if ((m == (1 << fmt.SigBits - 1)) &&
+            d.compareTo(BigInteger.ZERO) < 0 &&
```
```suggestion
            d.signum() < 0 &&
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      if (d2.compareTo(y) < 0) { // d2 < y
+        if ((m == (1 << fmt.SigBits - 1)) &&
+            d.compareTo(BigInteger.ZERO) < 0 &&
+            d2.multiply(BigInteger.TWO).compareTo(y) > 0)
```
```suggestion
            d2.shiftLeft(1).compareTo(y) > 0)
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      val d2 = BigInteger.TWO.multiply(mantissa).multiply(d.abs())
+
+      if (d2.compareTo(y) < 0) { // d2 < y
+        if ((m == (1 << fmt.SigBits - 1)) &&
```
Precedence of `<<` and `-` is not obvious at all here. Add parentheses around the right one (I honestly don't know which one it is).

tanishiking replied:
:+1: added parentheses around it to `if ((m == (1 << (fmt.SigBits - 1))` and added some explanations.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+            d2.multiply(BigInteger.TWO).compareTo(y) > 0)
+          loop(fmt.prevFloat(z))
+        else z
+      } else if (d2.compareTo(y) == 0) { // on boundary, round to even
```
Another comparison done twice.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+      } else if (d2.compareTo(y) == 0) { // on boundary, round to even
+        if (m % 2 == 0) {
+          if (m == (1 << Binary32.SigBits - 1) &&
+              d.compareTo(BigInteger.ZERO) < 0) loop(fmt.prevFloat(z))
```
```suggestion
              d.signum() < 0) loop(fmt.prevFloat(z))
```

sjrd replied:
And 2 more like that below.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+    }
+    loop(fmt.toIEEE754(z0))
+  }
+}
```
Missing EOL at EOF.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala:
```diff
@@ -0,0 +1,288 @@
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang.dec2flt
```
I'm not a fan of digits in package names. Consider `dectoflt`.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+import java.math.BigInteger
+import java.lang.{Long => jlLong}
+
+/** DIY floating point number with 64-bit significand bits */
```
```suggestion
/** DIY floating point number with 64-bit significant bits */
```
(with a 't' if used as an adjective; with a 'd' if used as a noun)

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+package java.lang.dec2flt
+
+import java.math.BigInteger
+import java.lang.{Long => jlLong}
```
Our convention for this case is `JLong`.
```suggestion
import java.lang.{Long => JLong}
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+    val halfway = 1L << (excessBits - 1)
+
+    val newF = {
+      if (jlLong.compareUnsigned(remainder, halfway) > 0) // roundup
```
Now you can use
```suggestion
      if (jlLong.unsigned_>(remainder, halfway)) // roundup
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+    val newF = {
+      if (jlLong.compareUnsigned(remainder, halfway) > 0) // roundup
+        shiftedF + 1
+      else if (jlLong.compareUnsigned(remainder, halfway) < 0)
```
and
```suggestion
      else if (jlLong.unsigned_<(remainder, halfway))
```
which are faster.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+        shiftedF + 1
+      else if (jlLong.compareUnsigned(remainder, halfway) < 0)
+        shiftedF
+      else if (shiftedF % 2 == 1) // tie, rounddown is odd
```
`shiftF` is interpreted as an unsigned quantity. We cannot meaningfully use signed `%` on it.
```suggestion
      else if ((shiftedF & 1) == 1) // tie, rounddown is odd
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+    val (sig, exponent) = roundNormal(fmt)
+
+    // Remove the leading implicit bit
+    val encodedSig: Long = sig - (1 << (fmt.ExplicitSigBits))
```
This is suspicious. Should it be
```suggestion
    val encodedSig: Long = sig - (1L << (fmt.ExplicitSigBits))
```
?

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+      shiftedF + 1
+    } else if (remainder < halfway) { // rounddown
+      shiftedF
+    } else if (shiftedF % 2 == 1) { // tie, rownddown (shiftedF) is odd (1 at LSB)
```
```suggestion
    } else if ((shiftedF & 1) == 1) { // tie, rownddown (shiftedF) is odd (1 at LSB)
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+}
+
+object FloatingPoint {
+  private val MASK: Long = 0xffffffffL
```
Unfortunately this makes it not a constant in Scala. Write as
```suggestion
  private final val MASK = 0xffffffffL
```
(with `final` and without `: Long`)

(also name it `Mask` to follow Scala conventions?)

tanishiking replied:
Right, and it seems this is unused 😁 removed.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+object FloatingPoint {
+  private val MASK: Long = 0xffffffffL
+
+  private val SigBits = 64
```
```suggestion
  private final val SigBits = 64
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+  private def normalize(f: BigInteger, e: Int): (Long, Int) = {
+    val shift = f.bitLength() - SigBits
+    if (shift == 0) (f.longValue(), e)
+    else if (shift < 0) {
```
```suggestion
    if (shift == 0) {
      (f.longValue(), e)
    } else if (shift < 0) {
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+
+  /** Normalize the given floating point number z = f * 2^e
+   *
+   *  A floating point number is normalized iff beta^(n-1) <= f < beta^n (beta=2, n=64)
```
Consider using the actual character β instead of `beta`. That's easier to understand in formulas.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+      val shiftedF = f.shiftRight(shift)
+      val remainder = f.and(BigInteger.ZERO.setBit(shift).subtract(BigInteger.ONE))
+      val halfway = BigInteger.ZERO.setBit(shift - 1)
+      val normalizedF =
```
```suggestion
      val normalizedF = {
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+  }
+}
```
Missing EOL

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+  type Repr = Float
+
+  final val ExpBits = 8
+  // final val ExplicitSigBits = 23
```
Remove?

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+  // final val ExplicitSigBits = 23
+  final val SigBits = 24
+
+  final val MaxSig: Long = (1L << SigBits) - 1;
```
```suggestion
  final val MaxSig = (1L << SigBits) - 1
```
otherwise it's not a real constant

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+
+  final val MaxSig: Long = (1L << SigBits) - 1;
+  final val CeilLog5OfMaxSig = 11
+  final val PositiveInfinity: Repr = java.lang.Float.POSITIVE_INFINITY
```
```suggestion
  final val PositiveInfinity = Float.PositiveInfinity
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+    val bits = java.lang.Float.floatToRawIntBits(v)
+    val m = (bits & 0x7fffff + (1 << ExplicitSigBits)).toLong
+    val exp = ((bits >>> ExplicitSigBits) & 0xff) -
+        ((1 << (ExpBits - 1)) - 1) - ExplicitSigBits
```
This reads really strangely. Keep it on one line:
```suggestion
    val exp = ((bits >>> ExplicitSigBits) & 0xff) - ((1 << (ExpBits - 1)) - 1) - ExplicitSigBits
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+
+  def powerOfTen(i: Int): Repr = PowerOfTens(i)
+  def prevFloat(v: Repr): Repr = Math.nextDown(v)
+  def nextFloat(v: Repr): Repr = Math.nextUp(v)
```
```suggestion
  def nextDown(v: Repr): Repr = Math.nextDown(v)
  def nextUp(v: Repr): Repr = Math.nextUp(v)
```
?

sjrd commented on javalib/src/main/scala/java/lang/dectoflt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+  def nextFloat(v: Repr): Repr
+  def mul(a: Repr, b: Repr): Repr
+  def div(a: Repr, b: Repr): Repr
+  def frexp(v: Repr): (Long, Int)
```
This one deserves a doc comment.

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+    val exp = ((bits >>> ExplicitSigBits) & 0xff) -
+        ((1 << (ExpBits - 1)) - 1) - ExplicitSigBits
+    (m, exp)
+  }
+
+  def reinterpretBits(v: BigInteger): Repr =
+    java.lang.Float.intBitsToFloat(v.intValue())
+  def fromLong(v: Long): Repr = v.toFloat
+  def toIEEE754(v: FloatingPoint): Repr = {
```
```suggestion
  def reinterpretBits(v: BigInteger): Repr =
    java.lang.Float.intBitsToFloat(v.intValue())

  def fromLong(v: Long): Repr = v.toFloat

  def toIEEE754(v: FloatingPoint): Repr = {
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/FloatingPointFormat.scala:
```diff
@@ -0,0 +1,154 @@
+    // Adjust the exponent for exponent bias and mantissa shift
+    val encodedExp: Int = exponent + ((1 << (ExpBits - 1)) - 1) + ExplicitSigBits
+    // combine bits
+    val bits = encodedExp << ExplicitSigBits | encodedSig
```
```suggestion
    val bits = (encodedExp << ExplicitSigBits) | encodedSig
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Tables.scala:
```diff
@@ -0,0 +1,192 @@
+    FP.normalized(-1634561335591402499L, 946), // 304 (19*16)
+    FP.normalized(-9114107888677362826L, 1000), // 320 (20*16)
+  )
+  val LargeTableRange = 20
```
```suggestion
  final val LargeTableRange = 20
```

sjrd commented on javalib/src/main/scala/java/lang/dec2flt/Tables.scala:
```diff
@@ -0,0 +1,192 @@
+    FP.normalized(-9114107888677362826L, 1000), // 320 (20*16)
+  )
+  val LargeTableRange = 20
+}
```
Missing EOL

sjrd commented on javalib/src/main/scala/java/math/BigInteger.scala:
```diff
@@ -42,6 +42,8 @@
 
 import scala.annotation.tailrec
+import scala.scalajs.LinkingInfo.linkTimeIf
+import scala.scalajs.LinkingInfo.targetPureWasm
```
Unused.

sjrd commented on javalib/src/main/scala/java/math/BigInteger.scala:
```diff
@@ -51,6 +53,8 @@ object BigInteger {
 
   final val ONE = new BigInteger(1, 1)
 
+  final val TWO = new BigInteger(1, 2)
```
I think after all my comments, this is not needed anymore.

(It's a real JDK API which could be submitted upstream as a separate PR, but there's no reason to add it here.)

sjrd commented on javalib/src/main/scala/java/util/regex/RegExpImpl.scala:
```diff
@@ -0,0 +1,73 @@
+
+object RegExpImpl {
+  // TODO: for debug, linkTimeIf(targetPureWasm)
+  val impl = linkTimeIf[RegExpImpl](isWebAssembly) {
```
Revert the debug?

sjrd commented on javalib/src/main/scala/java/util/regex/RegExpImpl.scala:
```diff
@@ -0,0 +1,73 @@
+  def getOrElse(r: Repr, index: Int, default: String): String
+}
+
+object RegExpImpl {
```
```suggestion
private[java] object RegExpImpl {
```

sjrd commented on javalib/src/main/scala/java/util/regex/RegExpImpl.scala:
```diff
@@ -0,0 +1,73 @@
+  def compile(patternStr: String): PatRepr
+  def exec(pattern: PatRepr, string: String): Repr
+  def matches(r: Repr): scala.Boolean
+  def exists(r: Repr, index: Int): scala.Boolean
```
The `scala.` prefixes are not necessary once we're outside the `java.lang` package.

sjrd commented on javalib/src/main/scala/java/util/regex/RegExpImpl.scala:
```diff
@@ -0,0 +1,73 @@
+      if (result != null) result else default
+    }
+  }
+}
```
Missing EOL

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -349,14 +346,61 @@ class DoubleTest {
+    // test("1e400", Double.PositiveInfinity);
+    test("1e309", Double.PositiveInfinity);
+    test("2e308", Double.PositiveInfinity);
+    // test("1.7976931348624e308", Double.PositiveInfinity);
```
TODO's?

tanishiking replied:
Oops, I forgot to remove `//` after debugging 😅 
And it turned out there's a problem in overflow detection when rounding from 64-bit extended precision floating point number to Float or Double, fixed in https://github.com/scala-wasm/scala-wasm/pull/106/commits/870898dfa42b74bfcde0c8c2cccaae549989d598

sjrd reviewed (changes requested):
Looks good overall. I only have local comments at this point.

I did not review the logic of the Bellerophon algorithm; I did not match it up against the paper. However, I did a thorough review of the details, and how the computations can be improved.

tanishiking commented:
Thanks a lot @sjrd for taking a look!
I think addressed all feedbacks in https://github.com/scala-wasm/scala-wasm/pull/106/commits/47026474cb31e6d59ab972995e8f24bd01c31cf0 and https://github.com/scala-wasm/scala-wasm/pull/106/commits/870898dfa42b74bfcde0c8c2cccaae549989d598, could you take an another look everything is fixed?

sjrd commented on javalib/src/main/scala/java/math/BigInteger.scala:
```diff
@@ -42,6 +42,7 @@
 package java.math
 
 import scala.annotation.tailrec
+import scala.scalajs.LinkingInfo.linkTimeIf
```
Leftover.

sjrd reviewed (approved):
LGTM :)
